rust/src/fvm/machine.rs

```diff
@@ -0,0 +1,264 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::mem;
+use std::ptr;
+use std::sync::{atomic::AtomicU64, Mutex};
+
+use anyhow::Error;
+use blockstore::cgo::CgoBlockstore;
+use blockstore::{Block, Blockstore, MemoryBlockstore};
+use cid::Cid;
+use drop_struct_macro_derive::DropStructMacro;
+use ffi_toolkit::{
+    c_str_to_pbuf, catch_panic_response, raw_ptr, rust_str_to_c_str, FCPResponseStatus,
+};
+use fvm::externs::cgo::CgoExterns;
+use fvm::externs::Externs;
+use fvm::machine::{ApplyKind, ApplyRet, Machine};
+use fvm::message::Message;
+use fvm::Config;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::encoding::RawBytes;
+use fvm_shared::version::NetworkVersion;
+use fvm_shared::MethodNum;
+use log::{error, info};
+use num_traits::FromPrimitive;
+use once_cell::sync::Lazy;
+
+use super::types::*;
+use crate::util::api::init_log;
+
+type CgoMachine = Machine<CgoBlockstore, CgoExterns>;
+
+static FVM_MAP: Lazy<Mutex<HashMap<u64, CgoMachine>>> =
+    Lazy::new(|| Mutex::new(HashMap::with_capacity(1)));
+
+const NEXT_ID: AtomicU64 = AtomicU64::new(0);
+
+fn add_fvm_machine(machine: CgoMachine) -> u64 {
+    let next_id = NEXT_ID.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+    let mut machines = FVM_MAP.lock().unwrap();
+    machines.insert(next_id, machine);
+    next_id
+}
+
+fn get_default_config() -> fvm::Config {
+    Config {
+        initial_pages: 1024, //FIXME
+        max_pages: 32768,    // FIXME
```

Comment:
@Stebalien any sense of the values we want to set here? Noting that WASM page size is 64KiB, so we may want to start with 16 pages (1MiB), and have no upper limit for M0? We probaly want to instrument the WASM runtime to record the pages we ended up allocating with every execution.

Reply:
No upper limit for now. We probably also need an exception for cron.

And yeah, we need to measure it. Then 10x at least.

Reply:
Note: I'd actually base the max memory on the total callstack. So we'll configure some max pages here, then set max_pages to `max_pages - used_pages` when instantiating a new instance.